File: test/samples/keywords/keywords.js
"use strict";

/**
 * @param {number} if_ not a condition
 * @param {string} class_ not a class
 * @param {string} i just a string
 * @param {{a: number, static: number}} in_ not in
 * @param {Array.<number>} for_ not a loop
 * @param {Array.<{int: {return: number, void: Array.<number>}, 'if true': number}>} words contains lots of things
 * @param {number} words1 an integer
 * @returns {void}
 */
function keywords(if_, class_, i, in_, for_, words, words1) {
    /* TODO If this compiles, it is already a good step! */
}

function main(stdin) {
    let line = 0;

    const if_ = Number(stdin[line++]);
    const class_ = stdin[line++];
    const i = stdin[line++];
    const words2 = stdin[line++].split(" ");
    const in_ = {
        a: Number(words2[0]),
        static_: Number(words2[1])
    };
    const for_ = stdin[line++].split(" ", if_).map(Number);
    const words = [];
    for (let j = 0; j < 2; j++) {
        const k = {};
        k.int_ = {};
        k.int_.return_ = Number(stdin[line++]);
        k.int_.void_ = stdin[line++].split(" ", 3).map(Number);
        k.ifTrue = Number(stdin[line++]);
        words.push(k);
    }
    const words1 = Number(stdin[line++]);
    keywords(if_, class_, i, in_, for_, words, words1);
}

let stdin = "";
process.stdin.on("data", data => stdin += data.toString())
             .on("end", () => main(stdin.split("\n")));

File: test/samples/keywords/keywords.ml
(* Emulate List.init from OCaml 4.06 *)
module List = struct
  include List

  let init n f =
    let rec aux i =
      if i >= n then [] else
        let r = f i in
        r :: aux (i+1) in
    aux 0
end

(* Copy String.split_on_char from OCaml 4.04 *)
module String = struct
  include String

  let split_on_char sep s =
    let r = ref [] in
    let j = ref (String.length s) in
    for i = String.length s - 1 downto 0 do
      if String.unsafe_get s i = sep then begin
        r := String.sub s (i + 1) (!j - i - 1) :: !r;
        j := i
      end
    done;
    String.sub s 0 !j :: !r
end

(** may conflict in c# *)
type console = {
  a : int; (** the first letter of the alphabet *)
  static : int; (** an integer *)
}

(** may conflict in c# *)
type system = {
  return : int; (** not the end of the function *)
  void : int list; (** not nothing *)
}

(** not the main function *)
type main = {
  int : system; (** not an integer *)
  ifTrue : int; (** should not cause conflict *)
}

(**
   @param if_ not a condition
   @param class_ not a class
   @param i just a string
   @param in_ not in
   @param for_ not a loop
   @param words contains lots of things
*)
let keywords if_ class_ i in_ for_ words =
  (** TODO If this compiles, it is already a good step! *)
  ()

let () =
  let if_ = read_int () in
  let class_ = (read_line ()).[0] in
  let i = read_line () in
  let in_ = Scanf.sscanf (read_line ()) "%d %d" (fun a static -> {a; static}) in
  let for_ = read_line () |> fun x -> if x = "" then [] else String.split_on_char ' ' x |> List.rev_map int_of_string |> List.rev in
  let words = List.init 2 (fun _ -> let int = let return = read_int () in let void = read_line () |> fun x -> if x = "" then [] else String.split_on_char ' ' x |> List.rev_map int_of_string |> List.rev in {return; void} in let ifTrue = read_int () in {int; ifTrue}) in
  keywords if_ class_ i in_ for_ words
